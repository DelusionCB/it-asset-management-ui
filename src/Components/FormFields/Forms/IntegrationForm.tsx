import React, {useEffect, useState} from 'react';
import {Row} from 'reactstrap';
import {formTypes} from '../../../Types/types.forms';
import VisibilitySelector from '../../CustomComponents/VisibilitySelector';
import {integrationDataPropTypes} from '../../../Types/types.directories';
import {integrationDefaultValues} from '../../../Types/defaultValues';
import {useTranslation} from 'react-i18next';
import {
    debouncedCheck,
    handleClear,
    handleIntegrationChange, handleIntegrationSelect,
} from '../../../Utils/helpers';
import TextField from '../../CustomComponents/TextField';
import StaticSelect from '../../CustomComponents/StaticSelect';
import ActionButtons from '../../ActionButtons/ActionButtons';
import SearchableSelect from '../../CustomComponents/SearchableSelect';

function IntegrationForm ({isDisabled, type, mode, editValues}: formTypes): JSX.Element {
    const [values, setValues] = useState<integrationDataPropTypes>(integrationDefaultValues)
    const {t} = useTranslation()

    function handleChange (e: React.ChangeEvent<HTMLInputElement>): void {
        handleIntegrationChange(e, values, setValues);
    }

    function handleSelect (value: {} | [] | undefined | null, id: string): void {
        handleIntegrationSelect(value, id, values, setValues);
    }

    useEffect(() => {
        if (Object.keys(editValues).length !== 0 && (mode === 'edit' || mode === 'copy')) {
            setValues(editValues)
        }
        if (Object.keys(editValues).length !== 0 && mode === 'create') {
            setValues(integrationDefaultValues)
        }
    }, [editValues, mode])

    useEffect(() => {
        if (mode !== 'edit') {
            // Invoke debouncedCheck with updated values
            debouncedCheck(values, integrationDefaultValues, isDisabled);

            // Cleanup function to cancel debounce on unmount
            return () => {
                debouncedCheck.cancel();
            };
        }
    }, [values]);

    useEffect(() => {
        if (values.visibility === 'hidden' || values.visibility === 'draft') {
            // Clear these fields if visibility is X -value
            handleClear<integrationDataPropTypes>(
                values,
                ['server_platform'],
                [],
                setValues,
            )
        }
    }, [values.visibility])

    return (
        <>
            <Row>
                <h1>{t(`integration.${mode}`)}</h1>
            </Row>
            <Row>
                <h2>{t('integration.visibility')}</h2>
            </Row>
            <Row>
                <VisibilitySelector
                    value={values.visibility}
                    onChange={(e) => { handleChange(e); }}
                    disabled={false}
                    id='visibility'
                    label={'values.visibility'}
                    invalid={false}
                    options={['draft', 'hidden', 'published']}
                    placeholder={'placeholder.visibility'}
                />
            </Row>
            <Row>
                <h2>{t('application.description')}</h2>
            </Row>
            <Row>
                <TextField
                    id='name'
                    disabled={false}
                    invalid={false}
                    label={'integration.values.name'}
                    onChange={(e) => { handleChange(e); }}
                    placeholder={'placeholder.name'}
                    value={values.name}
                    type='text'
                />
            </Row>
            <Row>
                <TextField
                    id={'description'}
                    disabled={false}
                    invalid={false}
                    type='textarea'
                    label={'integration.values.description'}
                    onChange={(e) => { handleChange(e) }}
                    placeholder={'placeholder.description'}
                    value={values.description}
                />
            </Row>
            <Row>
                <SearchableSelect
                    label={'integration.values.server_platform'}
                    endpoint='server'
                    id={'server_platform'}
                    placeholder={'placeholder.server_platform'}
                    disabled={values.visibility === 'hidden' || values.visibility === 'draft'}
                    invalid={false}
                    isMulti={false}
                    value={values.server_platform}
                    onChange={(value, id) => { handleSelect(value, id); }}
                />
            </Row>
            <Row>
                <StaticSelect
                    id={'environment_type'}
                    disabled={false}
                    invalid={false}
                    label={'integration.values.environment_type'}
                    onChange={(e) => { handleChange(e) }}
                    placeholder={'placeholder.environment_type'}
                    value={values.environment_type}
                    options={['dev', 'test', 'stage', 'prod']}
                />
            </Row>
            <ActionButtons
                values={values}
                type={type}
                mode={mode}
            />
        </>
    );
}

export default IntegrationForm;
