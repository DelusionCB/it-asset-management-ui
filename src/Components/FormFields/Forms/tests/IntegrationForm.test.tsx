import React from 'react';
import * as TestRenderer from 'react-test-renderer';
import {BrowserRouter} from 'react-router-dom'
import '@testing-library/jest-dom'
import IntegrationForm from '../IntegrationForm';

describe('<IntegrationForm />', () => {
    let renderer: TestRenderer.ReactTestRenderer;

    // Render the component before tests
    beforeAll(() => {
        renderer = TestRenderer.create(
            <BrowserRouter>
                <IntegrationForm editValues={{}} mode='create' type='integration' isDisabled={jest.fn()} />
            </BrowserRouter>,
        );
    });
    describe('Snapshot', () => {
        it('renders correctly', () => {
            expect(renderer.toJSON()).toMatchSnapshot();
        })
    })
})
