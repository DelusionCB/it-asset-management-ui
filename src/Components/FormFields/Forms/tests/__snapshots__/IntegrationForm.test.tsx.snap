// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IntegrationForm /> Snapshot renders correctly 1`] = `
[
  <div
    className="row"
  >
    <h1>
      integration.create
    </h1>
  </div>,
  <div
    className="row"
  >
    <h2>
      integration.visibility
    </h2>
  </div>,
  <div
    className="row"
  >
    <div
      className="textfield col"
    >
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="visibility"
        >
          values.visibility
        </label>
        <select
          aria-invalid={false}
          className="form-select"
          disabled={false}
          id="visibility"
          name="visibility"
          onChange={[Function]}
          placeholder="placeholder.visibility"
          value="published"
        >
          <option
            value="draft"
          >
            visibility.draft
          </option>
          <option
            value="hidden"
          >
            visibility.hidden
          </option>
          <option
            value="published"
          >
            visibility.published
          </option>
        </select>
      </div>
    </div>
  </div>,
  <div
    className="row"
  >
    <h2>
      application.description
    </h2>
  </div>,
  <div
    className="row"
  >
    <div
      className="textfield col"
    >
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="name"
        >
          integration.values.name
        </label>
        <input
          aria-invalid={false}
          className="form-control"
          disabled={false}
          id="name"
          name="name"
          onChange={[Function]}
          placeholder="placeholder.name"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="textfield col"
    >
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="description"
        >
          integration.values.description
        </label>
        <textarea
          aria-invalid={false}
          className="form-control"
          disabled={false}
          id="description"
          name="description"
          onChange={[Function]}
          placeholder="placeholder.description"
          value=""
        />
      </div>
    </div>
  </div>,
  <div
    className="row"
  >
    <div>
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="server_platform"
        >
          integration.values.server_platform
        </label>
        <div
          className=" css-b62m3t-container"
          onKeyDown={[Function]}
        >
          <span
            className="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            className="css-1f43avz-a11yText-A11yText"
          />
          <div
            className="option-search__control css-13cymwt-control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="option-search__value-container css-1fdsijx-ValueContainer"
            >
              <div
                className="option-search__placeholder css-1jqq78o-placeholder"
                id="react-select-2-placeholder"
              >
                placeholder.server_platform
              </div>
              <div
                className="option-search__input-container css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-2-placeholder"
                  aria-expanded={false}
                  aria-haspopup={true}
                  aria-invalid={false}
                  aria-label="placeholder.server_platform"
                  aria-labelledby="server_platform"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  className="option-search__input"
                  disabled={false}
                  id="server_platform"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  role="combobox"
                  spellCheck="false"
                  style={
                    {
                      "background": 0,
                      "border": 0,
                      "color": "inherit",
                      "font": "inherit",
                      "gridArea": "1 / 2",
                      "label": "input",
                      "margin": 0,
                      "minWidth": "2px",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "100%",
                    }
                  }
                  tabIndex={0}
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              className="option-search__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                className="option-search__indicator-separator css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                className="option-search__indicator option-search__dropdown-indicator css-1xc3v61-indicatorContainer"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-tj5bde-Svg"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <input
            name="server_platform"
            type="hidden"
            value=""
          />
        </div>
      </div>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="textfield col"
    >
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="environment_type"
        >
          integration.values.environment_type
        </label>
        <select
          aria-invalid={false}
          className="form-select"
          disabled={false}
          id="environment_type"
          name="environment_type"
          onChange={[Function]}
          placeholder="placeholder.environment_type"
          value="prod"
        >
          <option
            value="dev"
          >
            select.dev
          </option>
          <option
            value="test"
          >
            select.test
          </option>
          <option
            value="stage"
          >
            select.stage
          </option>
          <option
            value="prod"
          >
            select.prod
          </option>
        </select>
      </div>
    </div>
  </div>,
  <div
    className="action-buttons"
  >
    <button
      className="action-button btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      button.cancel
    </button>
    <button
      className="action-button btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      button.create
    </button>
  </div>,
]
`;
